include/proxy-wasm/exports.h

```diff
@@ -153,5 +154,51 @@ Word pthread_equal(void *, Word left, Word right);
+                              _f(grpc_cancel) _f(grpc_send) _f(set_tick_period_milliseconds)       \
+                                  _f(get_current_time_nanoseconds) _f(define_metric)               \
+                                      _f(increment_metric) _f(record_metric) _f(get_metric)        \
+                                          _f(set_effective_context) _f(done)                       \
+                                              _f(call_foreign_function)
+
+#define FOR_ALL_HOST_FUNCTIONS_ABI_SPECIFIC(_f)                                                    \
+  _f(get_configuration) _f(continue_request) _f(continue_response) _f(clear_route_cache)           \
+      _f(continue_stream) _f(close_stream) _f(get_log_level)
+
+#define FOR_ALL_WASI_FUNCTIONS(_f)                                                                 \
+  _f(fd_write) _f(fd_read) _f(fd_seek) _f(fd_close) _f(fd_fdstat_get) _f(environ_get)              \
+      _f(environ_sizes_get) _f(args_get) _f(args_sizes_get) _f(clock_time_get) _f(random_get)      \
+          _f(proc_exit)
+
+// Helpers to generate a stub to pass to VM, in place of a restricted export.
+#define _CREATE_EXPORT_STUB(_fn)                                                                   \
+  template <typename F> struct _fn##Stub;                                                          \
+  template <typename... Args> struct _fn##Stub<Word(void *, Args...)> {                            \
+    static Word stub(void *raw_context, Args...) {                                                 \
+      auto context = exports::ContextOrEffectiveContext(                                           \
+          static_cast<ContextBase *>((void)raw_context, current_context_));                        \
+      context->wasmVm()->integration()->error("Attempted call to restricted capability: " #_fn);   \
+      return WasmResult::InternalFailure;                                                          \
+    }                                                                                              \
+  };                                                                                               \
+  template <typename... Args> struct _fn##Stub<void(void *, Args...)> {                            \
+    static void stub(void *raw_context, Args...) {                                                 \
+      auto context = exports::ContextOrEffectiveContext(                                           \
+          static_cast<ContextBase *>((void)raw_context, current_context_));                        \
+      context->wasmVm()->integration()->error("Attempted call to restricted capability: " #_fn);   \
+    }                                                                                              \
+  };
+FOR_ALL_HOST_FUNCTIONS(_CREATE_EXPORT_STUB)
+FOR_ALL_HOST_FUNCTIONS_ABI_SPECIFIC(_CREATE_EXPORT_STUB)
+FOR_ALL_WASI_FUNCTIONS(_CREATE_EXPORT_STUB)
```

Comment:
For WASI, I think it would be better if we had a dedicated stub that returns `__WASI_ENOTCAPABLE` (`76`), since WASI modules should be aware of that error code, and they know nothing about Proxy-Wasm return values.